src/dvc_data/objects/tree.py

```diff
@@ -267,26 +266,35 @@ def _diff(ancestor, other, allow_removed=False):
     return result
 
 
-def _merge(ancestor, our, their):
+def _merge(ancestor, our, their, allowed=None):
     import copy
 
-    from dictdiffer import patch
+    from dictdiffer import diff, patch
 
-    our_diff = _diff(ancestor, our)
+    our_diff = _diff(ancestor, our, allowed=allowed)
     if not our_diff:
         return copy.deepcopy(their)
 
-    their_diff = _diff(ancestor, their)
+    their_diff = _diff(ancestor, their, allowed=allowed)
     if not their_diff:
         return copy.deepcopy(our)
 
     # make sure there are no conflicting files
-    _diff(our, their, allow_removed=True)
-
-    return patch(our_diff + their_diff, ancestor)
+    patch_ours_first = patch(our_diff + their_diff, ancestor)
+    patch_theirs_first = patch(their_diff + our_diff, ancestor)
+    unmergeable = list(diff(patch_ours_first, patch_theirs_first))
+    if unmergeable:
+        unmergeable_paths = []
+        for paths in patch(unmergeable, {}):
+            unmergeable_paths.append(posixpath.join(*paths))
+        raise MergeError(
+            "unable to auto-merge the following paths:\n"
+            + "\n".join(unmergeable_paths)
+        )
+    return patch_ours_first
```

Comment:
Quick explanation here (happy to add more comments):
1. Compare merge between ours->theirs vs theirs->ours.
2. If there are differences, consider it unmergeable.
3. Raise error listing the paths that differed.